crates/bevy_render/src/render_resource/pipeline.rs

```diff
@@ -118,14 +124,107 @@ pub struct VertexState {
 }
 
 /// Describes how the vertex buffer is interpreted.
-#[derive(Clone, Debug, Hash, Eq, PartialEq)]
+#[derive(Clone, Debug, Default, Eq)]
 pub struct VertexBufferLayout {
     /// The stride, in bytes, between elements of this buffer.
     pub array_stride: BufferAddress,
     /// How often this vertex buffer is "stepped" forward.
     pub step_mode: VertexStepMode,
     /// The list of attributes which comprise a single vertex.
-    pub attributes: Vec<VertexAttribute>,
+    attributes: Vec<VertexAttributeLayout>,
+    /// The list of attributes suitable for `wgpu`.
+    wgpu_attributes: Vec<VertexAttribute>,
+    /// Attribute names for debugging and mapping types.
+    attribute_names: HashMap<String, usize>,
+}
+
+impl Hash for VertexBufferLayout {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.array_stride.hash(state);
+        self.step_mode.hash(state);
+        self.attributes.hash(state);
+    }
+}
+
+impl PartialEq for VertexBufferLayout {
+    fn eq(&self, other: &Self) -> bool {
+        self.array_stride == other.array_stride
+            && self.step_mode == other.step_mode
+            && self.attributes == other.attributes
+    }
+}
+
+impl VertexBufferLayout {
+    /// Push a vertex attribute descriptor to the end of the list.
+    ///
+    /// The shader location is determined based on insertion order.
+    pub fn push(&mut self, name: &str, format: VertexFormat) {
+        let shader_location = if let Some(attribute) = self.attributes.last() {
+            attribute.shader_location + 1
+        } else {
+            0
+        };
```

Comment:
This might be extremely nit-picky, and `clippy` could say as much, but this can be rewritten as following:
```suggestion
        let shader_location = self.attributes.last().map_or(0, |attr| attr.shader_location + 1);
```